include/dlaf/communication/sync/reduce.h

```diff
@@ -13,24 +13,102 @@
 /// @file
 
 #include <mpi.h>
+#include "dlaf/common/assert.h"
+#include "dlaf/common/data_descriptor.h"
 #include "dlaf/communication/communicator.h"
 #include "dlaf/communication/message.h"
 
 namespace dlaf {
 namespace comm {
 namespace sync {
 
-/// MPI_Reduce wrapper
+namespace internal {
+namespace reduce {
+
+/// @brief MPI_Reduce wrapper (sender side)
+/// MPI Reduce(see MPI documentation for additional info)
+/// @param reduce_operation MPI_Op to perform on @p input data coming from ranks in @p communicator
+template <class DataIn, class DataOut>
+void collector(Communicator& communicator, MPI_Op reduce_operation, const DataIn input,
+               const DataOut output) {
+  using T = std::remove_const_t<typename common::data_traits<DataIn>::element_t>;
+
+  // Wayout for single rank communicator, just copy data
+  if (communicator.size() == 1) {
+    common::copy(input, output);
+    return;
+  }
+
+  common::Buffer<T> tmp_mem_input;
+  common::Buffer<T> tmp_mem_output;
+
+  common::DataDescriptor<const T> tmp_in = input;
+  common::DataDescriptor<T> tmp_out = output;
+
+  // if input is not contiguous, copy it in a contiguous temporary buffer
+  if (!input.is_contiguous()) {
+    tmp_in = tmp_mem_input = common::create_temporary_buffer(tmp_in);
```

Comment:
This is unusual, threw me off for a second, is this supposed to mean that:

```
tmp_mem_input = common::create_temporary_buffer(tmp_in);
tmp_in = tmp_mem_input
```

Reply:
Yeah, I knew that it may had been "problematic".

I would a comment, split the assignment and use more self-explaining names to make it more clear.

You have to consider the two different types and functionalities:
- `tmp_in` is a `DataDescriptor`, it does not own any memory, it just describes a memory from a `Data` concept.
- `tmp_mem_in` is a `Buffer` that implements the `Data` concept, it owns the memory.

So, it is exactly as you said:
- create a temporary buffer `Buffer` and reference to it through `tmp_mem_input`
- make `tmp_in` pointing to the just created temporary buffer in `tmp_mem_input`